client/clientimpl.go

```diff
@@ -0,0 +1,348 @@
+package client
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/gorilla/websocket"
+
+	"github.com/open-telemetry/opamp-go/client/internal"
+	"github.com/open-telemetry/opamp-go/client/types"
+	"github.com/open-telemetry/opamp-go/internal/protobufs"
+)
+
+// Impl is a Client implementation.
+type Impl struct {
+	logger   types.Logger
+	settings StartSettings
+
+	// OpAMP Server URL.
+	url *url.URL
+
+	// HTTP request headers to use when connecting to OpAMP Server.
+	requestHeader http.Header
+
+	// Websocket dialer and connection.
+	dialer    websocket.Dialer
+	conn      *websocket.Conn
+	connMutex sync.RWMutex
+
+	// True is Start() is successful.
+	isStarted bool
+
+	// Context to use for background running go routines.
+	runContext context.Context
+	runCancel  context.CancelFunc
+
+	// True when stopping is in progress.
+	isStoppingFlag  bool
+	isStoppingMutex sync.RWMutex
+
+	// Indicates that the Client is fully stopped.
+	stoppedSignal chan struct{}
+
+	// The sender is responsible for sending portion of the OpAMP protocol.
+	sender *internal.Sender
+}
+
+var _ OpAMPClient = (*Impl)(nil)
+
+const retryAfterHTTPHeader = "Retry-After"
+
+func New(logger types.Logger) *Impl {
+	if logger == nil {
+		logger = &nopLogger{}
+	}
+
+	w := &Impl{
+		logger:        logger,
+		stoppedSignal: make(chan struct{}, 1),
+		sender:        internal.NewSender(logger),
+	}
+	return w
+}
+
+func (w *Impl) Start(settings StartSettings) error {
+	if w.isStarted {
+		return errors.New("already started")
+	}
+
+	w.settings = settings
+
+	var err error
+
+	// Prepare server connection settings.
+	w.url, err = url.Parse(w.settings.OpAMPServerURL)
+	if err != nil {
+		return err
+	}
+
+	w.dialer = *websocket.DefaultDialer
+
+	if w.settings.TLSConfig != nil {
+		w.url.Scheme = "wss"
+	}
+	w.dialer.TLSClientConfig = w.settings.TLSConfig
+
+	if w.settings.AuthorizationHeader != "" {
+		w.requestHeader = http.Header{}
+		w.requestHeader["Authorization"] = []string{w.settings.AuthorizationHeader}
+	}
+
+	// Prepare the first status report.
+	w.sender.UpdateStatus(
+		func(statusReport *protobufs.StatusReport) {
+			statusReport.AgentDescription = &protobufs.AgentDescription{
+				AgentType:    w.settings.AgentType,
+				AgentVersion: w.settings.AgentVersion,
+			}
+
+			statusReport.ServerProvidedAllAddonsHash = w.settings.LastServerProvidedAllAddonsHash
+		},
+	)
+
+	w.startConnectAndRun()
+
+	w.isStarted = true
+
+	return nil
+}
+
+func (w *Impl) Stop(ctx context.Context) error {
+	if !w.isStarted {
```

Comment:
Should we assume that Start/Stop are thread-safe? If so, perhaps `isStarted` should be guarded or atomic variable could be used here? (then, maybe it's an overkill)

Reply:
I will add docs to Client interface to clarify.

Reply:
Clarified.